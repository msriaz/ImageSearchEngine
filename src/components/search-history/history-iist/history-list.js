import React, {forwardRef} from 'react';
import {FlatList} from 'react-native';

import {HistoryItem} from './history-item';

/**
 * Image Item Component
 * @param {data} defines the data of images to render
 * @param {loadMore} defines the function to be called for loading more data
 * @param {rest} defines additional props to pass to FlatList
 */
export const HistoryList = ({data, onPressHistoryItem, rest}) => {
  const renderItem = ({item}) => (
    <HistoryItem
      title={item}
      key={item}
      onPressHistoryItem={onPressHistoryItem}
    />
  );

  return (
    <FlatList data={data} renderItem={renderItem} {...rest} />
  );
};
