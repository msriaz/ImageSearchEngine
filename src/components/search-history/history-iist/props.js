import PropTypes from 'prop-types';

export const propTypes = {
  searchQuery: PropTypes.string.isRequired,
  onPressHistoryItem: PropTypes.func.isRequired,
};

export const defaultProps = {};
