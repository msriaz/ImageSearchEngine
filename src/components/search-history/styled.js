import styled from 'styled-components/native';

export const StyledTextInput = styled.TextInput`
  border: 1px;
  border-color: ${({theme}) => theme.color.lightGray};
  width: ${({theme}) => theme.windowWidth * 0.75}px;
  padding: ${({theme}) => theme.spacing.small}px;
`;

export const StyledButton = styled.Text``;

export const StyledWrapper = styled.View`
  justify-content: space-between;
  margin: ${({theme}) => theme.spacing.small}px;
`;

export const StyledHeader = styled.Text`
  font-size: 18px;
  font-weight: bold;
  margin-bottom: ${({theme}) => theme.spacing.small}px;
`;

export const HeaderWrapper = styled.View`
  justify-content: space-between;
  margin: ${({theme}) => theme.spacing.small}px;
  flex-direction: row;
  align-items: center;
`;

export const StyledText = styled.Text`
  font-size: 12px;
  font-weight: 200;
  /* margin-left: ${({theme}) => theme.spacing.default}px; */
`;
