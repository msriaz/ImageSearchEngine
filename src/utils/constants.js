import {color as colors} from '../theme';
import {generateColorPalleteConstants} from './index';

export const APP_COLORS = generateColorPalleteConstants(colors);

export const STORAGE_KEYS = {
  HISTORY: 'history',
};
