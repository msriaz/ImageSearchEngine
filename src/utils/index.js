import {useEffect, useRef} from 'react';
import snakeCase from 'lodash/snakeCase';
/**
 *
 * @param {*} value
 * @returns previous state
 */
export const usePrevious = value => {
  const ref = useRef();
  useEffect(() => {
    ref.current = value;
  });
  return ref.current;
};

/**
 *
 * @param {*} imgData
 * @returns image url to render
 */
export const getImageUrl = imgData => {
  const {farm, server, id, secret} = imgData;
  return `https://farm${farm}.static.flickr.com/${server}/${id}_${secret}.jpg`;
};

/**
 * generate color consts from color pallette
 * @param {*} colors
 */
export const generateColorPalleteConstants = colors => {
  const colorsList = {};
  Object.keys(colors).forEach(colorKey => {
    colorsList[snakeCase(colorKey).toUpperCase()] = colors[colorKey];
  });
  return colorsList;
};

/**
 *
 * @param {*} funcName
 * @param {*} error
 */
export const reportError = (funcName, error) => {
  console.log(`Error in ${funcName}`, error);
};
