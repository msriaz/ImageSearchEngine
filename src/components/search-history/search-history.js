import React, {useEffect, useState} from 'react';
import {
  StyledHeader,
  StyledWrapper,
  StyledButton,
  HeaderWrapper,
  StyledText,
} from './styled';
import {load, clear} from '../../utils/storage';
import {STORAGE_KEYS} from '../../utils/constants';
import {reportError} from '../../utils';
import {HistoryList} from './history-iist';

export const SearchHistory = ({onPressHistoryItem, searchQuery}) => {
  const [history, setHistory] = useState([]);

  useEffect(() => {
    loadHistory();
  }, [searchQuery]);

  const loadHistory = async () => {
    try {
      const savedHistory = await load(STORAGE_KEYS.HISTORY);
      console.log('load more history');
      if (savedHistory) {
        setHistory(savedHistory);
      }
    } catch (error) {
      reportError('loadHistory', error);
    }
  };

  const onClearHistory = async () => {
    try {
      await clear();
      setHistory([]);
    } catch (error) {
      reportError('onClearHistory', error);
    }
  };

  return history?.length ? (
    <StyledWrapper>
      <HeaderWrapper>
        <StyledHeader>Search History</StyledHeader>
        <StyledButton onPress={onClearHistory}>Clear History</StyledButton>
      </HeaderWrapper>
      <HistoryList data={history} onPressHistoryItem={onPressHistoryItem} />
    </StyledWrapper>
  ) : (
    <StyledWrapper>
      <StyledText> No previous history found </StyledText>
    </StyledWrapper>
  );
};
