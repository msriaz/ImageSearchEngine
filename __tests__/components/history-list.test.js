import React from 'react';
import {HistoryList} from '../../src/components/search-history/history-iist';
import {render} from '../utils/theme-wrapper';

const onPressMock = jest.fn();
const mockData = ['cats', 'dogs', 'cows', 'goats'];

describe('History List Component', () => {
  it('should match the History List  snapshot', () => {
    const compTree = render(
      <HistoryList data={mockData} onPressHistoryItem={onPressMock} />,
    );
    expect(compTree.toJSON()).toMatchSnapshot();
  });
});
