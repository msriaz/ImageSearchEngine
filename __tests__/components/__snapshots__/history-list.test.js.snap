// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History List Component should match the History List  snapshot 1`] = `
<RCTScrollView
  data={
    Array [
      "cats",
      "dogs",
      "cows",
      "goats",
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={true}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#F4F4F4",
            "borderRadius": 30,
            "opacity": 1,
            "paddingBottom": 8,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 8,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 14,
                "fontWeight": "200",
              },
            ]
          }
        >
          cats
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#F4F4F4",
            "borderRadius": 30,
            "opacity": 1,
            "paddingBottom": 8,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 8,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 14,
                "fontWeight": "200",
              },
            ]
          }
        >
          dogs
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#F4F4F4",
            "borderRadius": 30,
            "opacity": 1,
            "paddingBottom": 8,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 8,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 14,
                "fontWeight": "200",
              },
            ]
          }
        >
          cows
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#F4F4F4",
            "borderRadius": 30,
            "opacity": 1,
            "paddingBottom": 8,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 8,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 14,
                "fontWeight": "200",
              },
            ]
          }
        >
          goats
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
